Improve error messages about peer disconnections [ECR-2154, ECR-2156]

ilammy opened the pull request:
Include the peer address into the error message so that we at least know who tries to connect to us.

Peer address is not readily available from the Noise session (as we have only IO stream of the TCP socket) so thread it as an additional argument.

ilammy commented on exonum/src/events/network.rs:
```diff
@@ -215,7 +215,7 @@ impl ConnectionsPool {
             let mut handshake_params = handshake_params.clone();
             handshake_params.set_remote_key(remote_public_key);
-            NoiseHandshake::initiator(&handshake_params).send(stream)
+            NoiseHandshake::initiator(&handshake_params, peer).send(stream)
```
Maybe we could put peer address into `handshake_params` to avoid changing function signature. But TBH this socket address has nothing to do with Noise handshake so I have some doubts about that.

DarkEld3r replied:
I don't see problems in changing the function signature. You can add a comment why this parameter is needed if it bothers you.

ilammy commented:
Pulled updates from master, resolved conflicts with recently merged PR #474.

ilammy commented:
I have realized that `failure` crate allows us to provide better error messages with context. I've updated the code so that it logs better messages with more meaningful context. It resolves another issue as well. Updated PR title accordingly.
